### ISSUE TITLE
Trigger Control queue stuck

### ISSUE THREAD

**Issue by sgraystar** [autoSteve/acMqtt#38]
Gday Steve,
I haven't updated the script in a while but changed recently and noticed this problem with updates to a trigger control group.

Device: 5500AC2
Version: 1.15.0, hw v1.1
Script version: https://github.com/autoSteve/acMqtt/commit/da64373c4bb5c4123e90ba4ebebd87bf917f6741

Add trigger control group with keywords: MQTT, select, lvl=0/255, sa=ZZZ

```
MQTT Send Receive 06.05.2024 11:09:56
* string: Queued 1 object with keyword MQTT for publication: 0/202/11
MQTT Send Receive 06.05.2024 11:09:56
* string: cudCBusTopics() completed in 0.020 seconds
MQTT Send Receive 06.05.2024 11:09:56
* string: Publishing homeassistant/select/cbus_mqtt_254_202_11/config as breakfast in area ZZZ
MQTT Send Receive 06.05.2024 11:09:56
* string: Publishing select cbus/read/0/202/11 to off (0)
MQTT Send Receive 06.05.2024 11:09:56
* string: Published 1 CBus discovery and current level topic
``` 

Set trigger control group level = 255

```
MQTT Send Receive 06.05.2024 11:11:44
* string: Not setting 0/202/11 to 0, same as previous value
``` 

Set trigger control group level = 0

```
MQTT Send Receive 06.05.2024 11:12:40
* string: Setting 0/202/11 to 255, previous=255
MQTT Send Receive 06.05.2024 11:12:40
* string: Publishing select cbus/read/0/202/11 to on (255)
```

**Comment by autoSteve** [comment]
In the branch 38-trigger-control-queue-stuck I have made some changes. These are not yet merged to master.

Firstly to logging. The messages about the previous level were wrong. Was this actually setting the trigger to the correct level, @sgraystar? It was for me, just that the log was off.

Secondly, in order to test this I needed to switch a set of buttons at my joint to a select. This did not work. So all the other changes are to create/update/delete code, so that it correctly disposes of old discovery topics in favour of new.

**Comment by autoSteve** [comment]
While testing this scenario I did encounter a nil value being passed to the localbus callback, which was odd. This was throwing an error log message. Did you see anything in the error log?

I have plugged that corner case, where the script now checks and puts a warning in the event log. (Oops, did not reference this issue, but it's in the same branch.)

**Comment by sgraystar** [comment]
Unfortunately no change. Error log is empty.
To be sure I did a reset/clean-up of script storage & reboot.
Changing the trigger level from the admin GUI & monitoring from MQTT Explorer I got the same logging as above.

Trying again with add new group, add tag names for 0/255, set to 0, add keywords.

Toggling the group on/off from the GUI it does not pick up the 255, MQTT Explorer shows no change, repeat entries in the log:
```
MQTT Send Receive 08.05.2024 09:14:43
* string: Not setting 0/202/9 to 0, same as previous value
```

**Comment by sgraystar** [comment]
also, all config items set to false under --[[ SET AS APPROPRIATE TO SUIT ENVIRONMENT --]]

**Comment by sgraystar** [comment]
Correction: new group behaves the same way as first post

**Comment by autoSteve** [comment]
I just managed to replicate this, but I don't know why it's happening yet.

The second time I worked through the exact same sequence and it worked as expected. Earlier attempts also worked as expected.

Could I get you to set up the keywords, have it discover, and then restart MQTT send receive and check whether it works. I am wondering if this is create/update/delete related.

**Comment by sgraystar** [comment]
No change.
Add new group, add tag names for 0/255, set to 0, add keywords, observe publish, stop/start script, set levels.

**Comment by sgraystar** [comment]
Trying a few different things, changing local busTimeout = 1 in case of storage race condition, no difference.
Anywhere else I could insert an artificial delay?

Added an event script and using switch instead of select, same result. MQTT Explorer matches published info.

```
MQTT Send Receive 08.05.2024 15:08:33
* string: Queued 1 object with keyword MQTT for publication: 0/202/25
MQTT Send Receive 08.05.2024 15:08:33
* string: cudCBusTopics() completed in 0.013 seconds
MQTT Send Receive 08.05.2024 15:08:33
* string: Publishing homeassistant/switch/cbus_mqtt_254_202_25/config as testing in area ZZZ
MQTT Send Receive 08.05.2024 15:08:33
* string: Publishing state and level cbus/read/0/202/25 to OFF/0
MQTT Send Receive 08.05.2024 15:08:33
* string: Published 1 CBus discovery and current level topic
MQTT Send Receive 08.05.2024 15:08:56
* string: Not setting 0/202/25 to 0, same as previous value
Event for 0/202/25 08.05.2024 15:08:56
* arg: 1
  * string: Event for 0/202/25
* arg: 2
  * table:
   [1]
    * string: event.getvalue()
   [2]
    * number: 255
MQTT Send Receive 08.05.2024 15:09:03
* string: cudCBusTopics() completed in 0.010 seconds
MQTT Send Receive 08.05.2024 15:09:33
* string: cudCBusTopics() completed in 0.023 seconds
MQTT Send Receive 08.05.2024 15:09:45
* string: Setting 0/202/25 to 255, previous=0
MQTT Send Receive 08.05.2024 15:09:45
* string: Publishing state and level cbus/read/0/202/25 to ON/255
Event for 0/202/25 08.05.2024 15:09:45
* arg: 1
  * string: Event for 0/202/25
* arg: 2
  * table:
   [1]
    * string: event.getvalue()
   [2]
    * number: 0
``` 

Restart script, picks up actual value

```
* string: Publishing homeassistant/switch/cbus_mqtt_254_202_25/config as testing in area ZZZ
* string: Publishing state and level cbus/read/0/202/25 to OFF/0
``` 

Set level to 255, not published

```
MQTT Send Receive 08.05.2024 15:17:56
* string: Not setting 0/202/25 to 0, same as previous value
Event for 0/202/25 08.05.2024 15:17:56
* arg: 1
  * string: Event for 0/202/25
* arg: 2
  * table:
   [1]
    * string: event.getvalue()
   [2]
    * number: 255
```

**Comment by autoSteve** [comment]
I think this may be a timing thing.

New commit changes how trigger messages are on the localbus are processed.

**Comment by sgraystar** [comment]
Yippideedoodah. Initial test as a switch is good.

**Comment by autoSteve** [comment]
Hey @sgraystar.

I've made a change to extend examining datahex for _all_ applications when getting an event on the localbus. This has meant removing the specific "hack"/exception for app 202/trigger, which fixed the timing thing that you uncovered.

I'd appreciate an independent test of this code. It should operate the same for you. Branch All-apps-localbus-value-from-datahex, not yet merged to master.

It should do 202 handsomely, and also 228/250/255. Lighting groups were always a special case given ramping, so are unchanged. Decoding for bool, text, uint32, int32 and float32 are implemented, which covers 99.9% of use cases I'd reckon. Date? Time? RGB? Pfft.

**Comment by sgraystar** [comment]
OK. I'll try it out.
FYI, after loading the new branch I checked the logs and saw this message from before the change, while using the main branch.
It appears a few times as a single event or as a pair. No obvious pattern to the timing from my perspective.

```
* string: Error processing outstanding MQTT messages: [string "Resident"]:154: bad argument #1 to 'pairs' (table expected, got nil)
```

**Comment by sgraystar** [comment]
So far ...

Trigger Control, add tags 0/255, add keywords: MQTT, select, lvl=0/255
Registration OK
Set level = 255
Set Level = 10, undefined level messes things up for next step
Set level = 0, not published
```
MQTT Send Receive 10.05.2024 10:20:06
* string: Setting 0/202/50 to 255, previous=0
MQTT Send Receive 10.05.2024 10:20:06
* string: Publishing select cbus/read/0/202/50 to on (255)
MQTT Send Receive 10.05.2024 10:20:22
* string: Not setting 0/202/50 to 255, same as previous value
MQTT Send Receive 10.05.2024 10:20:58
* string: Setting 0/202/50 to 10, previous=255
MQTT Send Receive 10.05.2024 10:20:58
* string: Publishing select cbus/read/0/202/50 to on (255)
``` 

Remove keywords, wait for removal, add keywords
Set level = 255, not published. Suggests data not entirely cleared by keyword removal.
```
* string: Publishing homeassistant/select/cbus_mqtt_254_202_50/config as test trigger in area no preferred area
* string: Publishing select cbus/read/0/202/50 to off (0)
* string: Published 1 CBus discovery and current level topic
* string: Not setting 0/202/50 to 0, same as previous value
```

**Comment by autoSteve** [comment]
Regarding ```Error processing outstanding MQTT messages```, the only time that hasMembers() is called there is relating to covers. Do you have blinds set up? This is called for ```OPEN```, ```CLOSE```, ```STOP```, and when ramping a cover.

**Comment by sgraystar** [comment]
I do have blinds, using lighting groups for non-cbus blinds via HA. Only really setting 0/255 although the HA automations allow for any level in between.
Keywords: MQTT, cover, noleveltranslate, sa=ZZZ, img=mdi:roller-shade

FWIW I only added the noleveltranslate yesterday and have not seen the error before.

**Comment by autoSteve** [comment]
Then the error is relating to a keyword I added not so long ago, or rather the lack of it. ```rate=n/n```. (This allows the script to periodically publish an estimated progress of travel while ramping a cover, so that sliders in HA visually approximate the 'openness'.) I have no blinds without this keyword.

This should default to ```{}```, so why it is ```nil``` in this case is currently a mystery. I'll take a further look, but I'll bet the error occurs when adjusting a blind, hence the lack of a pattern.

**Comment by sgraystar** [comment]
Yeah, the timing would have been around scheduler times to set blind scenes.

Also probably the same data cleanout problem as above.
Trigger control set keywords: MQTT, select, lvl=0/255
Remove keyords, wait for topic removal.
Add keywords, MQTT, switch
MQTT Explorer shows cbus read topic with select=off
Set level=255, not published
```
* string: Not setting 0/202/50 to 0, same as previous value
```

**Comment by sgraystar** [comment]
User Parameter Boolean, keywords: MQTT, switch
Always created/published as ON regrdless of true/false state.

**Comment by sgraystar** [comment]
> Then the error is relating to a keyword I added not so long ago, or rather the lack of it. `rate=n/n`. (This allows the script to periodically publish an estimated progress of travel while ramping a cover, so that sliders in HA visually approximate the 'openness'.) I have no blinds without this keyword.
> 
> This should default to `{}`, so why it is `nil` in this case is currently a mystery. I'll take a further look, but I'll bet the error occurs when adjusting a blind, hence the lack of a pattern.

then maybe due to me doing storage reset/clean-up during tests

**Comment by autoSteve** [comment]
I found the cover issue, and have fixed it. In short, by setting ```noleveltranslate``` the rate was set to ```nil``` instead of ```{}```.

**Comment by autoSteve** [comment]
> User Parameter Boolean, keywords: MQTT, switch

I never considered this configuration. The code presently does not support it. But it will...

**Comment by sgraystar** [comment]
There are so many combinations, and the cbus gui will let everything through or allow you to change data type after the topic creation, so some type of whitelisting combinations would be protective of your sanity.

**Comment by autoSteve** [comment]
Nah. I'd rather play whack-a-mole. 😂

> A question on buttons, since they are stateless should they suppress repetitive press messages?

Far simpler to allow repeats. Can you see a downside? If a scene is set then setting it again essentially does nothing, yeah?

**Comment by sgraystar** [comment]
sorry, aappears to be a furphy and you already manage the kill msg
```
* string: Setting 0/202/7 to -1, previous=2
```

**Comment by autoSteve** [comment]
> Trigger Control, add tags 0/255, add keywords: MQTT, select, lvl=0/255
> Registration OK
> Set level = 255
> Set Level = 10, undefined level messes things up for next step
> Set level = 0, not published

Hmmm. I can't replicate this.

When I set level ```10``` the value published to MQTT is ```255```. This is because ```local selectExact = true``` is set, which causes the next higher level to be published to HA. Then when I set the level to ```0``` it sets the select in HA to ```off```.

**Comment by sgraystar** [comment]
Trying to repeat I stumbled at the first step :(

Updated to latest commit https://github.com/autoSteve/acMqtt/commit/251b07278c0328e36887e0c47fbfa30436f16894
Reset storage & reboot
New trigger Control group, set level=0, add keywords: MQTT, select, lvl=0/255
All OK
```
MQTT Send Receive 10.05.2024 16:01:14
* string: Queued 1 object with keyword MQTT for publication: 0/202/61
MQTT Send Receive 10.05.2024 16:01:14
* string: cudCBusTopics() completed in 0.012 seconds
MQTT Send Receive 10.05.2024 16:01:14
* string: Publishing homeassistant/select/cbus_mqtt_254_202_61/config as testing in area no preferred area
MQTT Send Receive 10.05.2024 16:01:14
* string: Publishing select cbus/read/0/202/61 to off (0)
MQTT Send Receive 10.05.2024 16:01:14
* string: Published 1 CBus discovery and current level topic
``` 

Set level = 255, not published
```
MQTT Send Receive 10.05.2024 16:02:14
* string: Not setting 0/202/61 to 0, same as previous value
```

**Comment by autoSteve** [comment]
Again, can't replicate. Spin up the latest commit. It contains additional logging for datahex, and comparison variables.

**Comment by autoSteve** [comment]
Apologies. Commit bcbc011

**Comment by autoSteve** [comment]
> Set level = 10, then 0 published off correctly, so also not able to reproduce.

Hmmm. With ```local selectExact = true``` set it should have set it to ```255```! Something ain't quite right.

**Comment by sgraystar** [comment]
New trigger Control group, set level=0, add keywords: MQTT, select, lvl=0/255
All OK
```
MQTT Send Receive 10.05.2024 17:26:22
* string: Queued 1 object with keyword MQTT for publication: 0/202/64
MQTT Send Receive 10.05.2024 17:26:22
* string: cudCBusTopics() completed in 0.021 seconds
MQTT Send Receive 10.05.2024 17:26:22
* string: Publishing homeassistant/select/cbus_mqtt_254_202_64/config as testing64 in area no preferred area
MQTT Send Receive 10.05.2024 17:26:22
* string: Publishing select cbus/read/0/202/64 to off (0)
MQTT Send Receive 10.05.2024 17:26:22
* string: Published 1 CBus discovery and current level topic
``` 

Set level = 255, published OK
```
MQTT Send Receive 10.05.2024 17:27:04
* string: Setting 0/202/64 to 255, previous=0
MQTT Send Receive 10.05.2024 17:27:04
* string: Publishing select cbus/read/0/202/64 to on (255)
``` 

Set level = 0, not published
```
MQTT Send Receive 10.05.2024 17:28:31
* string: Not setting 0/202/64 to 255, same as previous value
MQTT Send Receive 10.05.2024 17:28:31
* string: Content of datahex: 0000. Type=8. pre=255.00000. comp=255.00000
```

**Comment by sgraystar** [comment]
repeat test
New trigger Control group, set level=0, add keywords: MQTT, select, lvl=0/255
Set level = 255, not published
```
MQTT Send Receive 10.05.2024 17:36:23
* string: cudCBusTopics() completed in 0.010 seconds
MQTT Send Receive 10.05.2024 17:36:53
* string: Queued 1 object with keyword MQTT for publication: 0/202/65
MQTT Send Receive 10.05.2024 17:36:53
* string: cudCBusTopics() completed in 0.027 seconds
MQTT Send Receive 10.05.2024 17:36:53
* string: Publishing homeassistant/select/cbus_mqtt_254_202_65/config as testing65 in area no preferred area
MQTT Send Receive 10.05.2024 17:36:53
* string: Publishing select cbus/read/0/202/65 to off (0)
MQTT Send Receive 10.05.2024 17:36:53
* string: Published 1 CBus discovery and current level topic
MQTT Send Receive 10.05.2024 17:37:03
* string: Not setting 0/202/65 to 0, same as previous value
MQTT Send Receive 10.05.2024 17:37:03
* string: Content of datahex: 00FF. Type=8. pre=0.00000. comp=0.00000
```

**Comment by sgraystar** [comment]
> > Set level = 10, then 0 published off correctly, so also not able to reproduce.
> 
> Hmmm. With `local selectExact = true` set it should have set it to `255`! Something ain't quite right.

Problem is my shorthand, level 10 published 255 as you expected, then level 0 published correctly, so all well behaved.

**Comment by autoSteve** [comment]
I do believe I know what is going on. Stand by for commit.next...

**Comment by sgraystar** [comment]
Excellent. Easily missed.
Still the most responsive helpdesk going around ;)

**Comment by autoSteve** [comment]
I reckon she's good to merge, @sgraystar. A bit of final optimisation done for good measure.

Thank you so much for your patience, and taking the time to get this right.

**Comment by autoSteve** [comment]
Hey @sgraystar.

I'm reopening this because I noticed today that one of my test switches had become out of sync. HA was showing switch on, CBus state was a boolean user parameter that was 'false'. Setting the switch to off in HA resulted in it immmediately bouncing back to on.

I have added a commit f5e55a1, which forces CBus set for anything other than a floating point value, even if the current value is the same.

Recommending that you update to avoid the out-of-sync situations. I'm still trying to get to the bottom of why it can occur.
